### DISCUSSION TITLE
Provide a package for type conversion

### DISCUSSION THREAD

**Discussion by nibra** [joomla-framework/application#104] [Ideas]
As discussed in [this comment](https://github.com/joomla/joomla-cms/discussions/37664#discussioncomment-2640701), it might be useful to add a package that deals with type conversions.

This discussion is used to evaluate the possible scope and other details of such a package.

**Comment by korenevskiy** [comment]
I took the courage and copied the function description from discussion. To close the previous discussion.
I hope @nibra will join the discussion as an experienced participant.

```PHP
function isTrue($value){
	if($value === false || $value === null || $value == 0 || $value == '' || $value == 'false' || $value == '0'
		 || $value == 'disabled' || $value == 'off' || $value == 'no' || $value == 'none')
		return false;
	return true;
}
```
This is an approximate writing of the function. But the need for such a feature in Joomla is obvious.
This function is needed to analyze the attribute values for XML.
each attribute must be repeatedly rechecked for all possible values. Despite the fact that the variants of values for all attributes are the same. The code becomes dirty, and repeated checks create a multiple little cognitive strain.
You constantly need to think about converting types to values when comparing, Which is very distracting from programming. 
And there is also no single rule for analyzing XML attributes. Even a young developer will now adapt to the analysis of this function.
The idea was to have a place where you can place this function. But

@nibra   I don't understand the purpose of casting types to other types.
It's a good idea, but it sounds like a fairy tale. Or you @nibra  have already used a specific package for casting types, and therefore you already have experience how it works.
Or we let's describe specific criteria convert types. 
-.Will be type conversion occur for XML and for the Database? The fact is that there are no problems with type conversion for XML, except for determining the true value for an attribute  in a comparison condition. 
-.The broader meaning of type conversion is in converting database value types to object properties and vice versa.

In the past, I have already written a type conversion function for the properties of objects and database fields.
https://github.com/joomla/joomla-cms/pull/35247/files 
The `TypeConvert` function is in the class Table.  (`function TypeConvert($value = null, $type = '', $tz = null)`)
From creating a typed class, it became clear to me. That it doesn't make sense to create a separate type conversion function without using a typed `Table` class at all. A type conversion for a database can only be created at the same time as a typed table.

This table class is under development. This class is fully operational, and supports full backward compatibility. But testing this class requires a large number of applicants. So I decided to change this `Table` class. In the future, I will create a `TableType` class that will inherit from the classic `Table` class. And it is necessary to make minimal changes to the classic `Table` class so that it can be the parent of the `TableType` class.

@nibra  Did you mean to create a type conversion package, at the same time to support type conversion for the database ?

**Reply by korenevskiy** [reply]
```PHP
    public static function attributeToHungarianNotation($attributeName) {
        $pos1 = strpos($attributeName, '_');
        $pos2 = strpos($attributeName, '-');
        if(empty($pos1) && empty($pos2))
            return $attributeName;
        
        if($pos1){ 
            $as = explode('_', $attributeName);
            $as = array_map(function($a){return ucfirst(strtolower($a));},$as);
            $attributeName = join('', $as);
        }
        if($pos2){ 
            $as = explode('-', $attributeName);
            $as = array_map(function($a){return ucfirst(strtolower($a));},$as);
            $attributeName = join('', $as);
        }
        return lcfirst($attributeName);
    }

echo attributeToHungarianNotation('translate-default'); // result: translateDefault;
```
I also had a problem with XML attribute names and object property names. The peculiarity is that attribute names are allowed to use underscores and dashes in names. But it is preferable to use only Hungarian Notation in the names of object properties. And multiple checks for the presence of all kinds of attributes for one property, as well as the problem described at the very beginning causes problems. For these purposes, I use one conversion function. Perhaps it will be useful in this case.

**Comment by nibra** [comment]
> @nibra Did you mean to create a type conversion package, at the same time to support type conversion for the database ?

For the time being, this discussion should be open-ended. To solve the Boolean problem, a class is needed that converts string representations into PHP types. However, it may make sense to offer additional conversion classes for other targets, e.g. SQL, perhaps in a later step.

**Comment by korenevskiy** [comment]
I fully support you that SQL type conversion is not applicable in this topic, since it does not make sense without a typed table. We'll put it off for later.
Let's figure it out here. Do you propose to do a type conversion for each individual field property? For example: if it is a translateDefault attribute, then we will automatically get the Bool type, but if it is a Label attribute, then we should automatically get the string type.
But there is another option, you can convert types based on the values themselves. If the attribute value is 1, then its type will then be an integer, but if the value is TRUE, then the type conversion will be BOOL. 
In this case, we cannot know what value will be in the attribute of the XML element for earlier. So at the output we will get several different but similar values. But even in this case we have an analog of type conversion (bool), (int), (string)
For example, type conversion is based on the values themselves. But then it is not clear to what type we will convert the string with the value "1". Such a value can be as a number, but such a value can still remain a string. For example, if the XML attribute Label="1". We will see the same thing with Label="123" and Title="123".

I propose to divide the current task into 2 independent processes. Note that it is the independent processes.
1.Choosing the location of the file with functions, classes. 
2. Thinking through the logic and defining the functions in this file.

Later, we may move something and rename it. But now we can make a preliminary (draft) PR, 2-3 functions. This will already be the beginning for discussing the logic of the file. But also at the same time, beginners who do not understand what is at stake can familiarize themselves with the existing library and make their suggestions.

The method written above **`isTrue()`** , performs the function only when cast to the **BOOL** type. Because it is this type that causes problems when cast to a type. Later we can add the **`toInt()`**, **`toString()`** and other functions. Now let's define a place and a class name  for these functions.
**`\libraries\vendor\joomla\utilities\src\TypeHelper.php`** 
Is it possible here, do you think?

Can we use **`isTrue()`** functions without a class, but using with a namespace?
Example: 
**`namespace \Joomla\Type;`**
or
**`namespace Joomla\Utilities;`**

**Reply by nibra** [reply]
I think that the proposed tool can live in the Joomla\Utilities namespace. I'd simply call it 'Cast', f.x.
```php
namespace Joomla\Utilities;

class Cast
{
    public static function toBoolean($value): bool
    {
        // Implementation here

        return $result;
    }
}
```

On the other hand, it can be seen as a Formatter (instead of Cast) as well. If we do so, we open up for things like
`116` <-> `one hundred and sixteen`
`2022` <-> `MMXXII`
`👍` <-> `:thumbsup:`

An other approach would be to provide classes for scalar types.
```php
namespace Joomla\Type;

class Boolean
{
    private bool $value;

    public function __construct($value)
    {
        $this->__set('value', $value);
    }

    public function __set(string $property, $value): void
    {
        if ($property !== 'value') {
            throw new UnexpectedPropertyException($property);
        }

        // Implementation here

        $this->value = $result;
    }

    public function __get(string $property): bool
    {
        if ($property !== 'value') {
            throw new UnexpectedPropertyException($property);
        }

        return $this->value;
    }
}
```
This class could even remember the original value and use that knowledge in a `__toString()` method to provide the correct string for output on request.


> Can we use isTrue() functions without a class, but using with a namespace?

Technically, that's possible, but we never provided such convenience proxy functions in Joomla before, IIRC.

**Reply by korenevskiy** [reply]
But OK. I have just tested the possibility of the toString() function. I was comparing 2 different Boolean objects in a condition. Everything works great. The function can return an integer or a Boolean type. After the comparison, the condition compares these returned values, and we get the result of different and identical values.
```PHP
class Boolean {
    private bool $value;
    public function __construct($value) {
        $this->value = $value;
    }

    public function __toString(string $property): bool     { 
        return (int)$this->value;
    }
}
echo (new Boolean(True)) == (new Boolean(True)) ? 'True' : 'False'; // Result: True
echo '<br>';
echo (new Boolean(False)) == (new Boolean(True)) ? 'True' : 'False';  // Result: False
```
Let's define scenarios for the use of such objects. And we can also appeal to those who may have a need for this class.

> Technically, that's possible, but we never provided such convenience proxy functions in Joomla before, IIRC.

I suggest using the **`isTrue()`** function without a class, specifying just a namespace. The basis of programming is to write algorithms, any algorithm consists of a condition. I equate the importance of this function with the importance of the **`DIRECTORY_SEPARATOR`** constant.

**Reply by korenevskiy** [reply]
```
116 <-> one hundred and sixteen
2022 <-> MMXXII
👍 <-> :thumbsup:
```
`Cast::toBoolean();`
Please write an example below how you came up with this example.

**Reply by nibra** [reply]
> I equate the importance of this function with the importance of the DIRECTORY_SEPARATOR constant.

... which is not needed at all in PHP. That's why we dropped it's usage years ago in most places.

**Reply by nibra** [reply]
> Please write an example below how you came up with this example.

If the solution was a Formatter class, it could perform a lot of different formating tasks, f.x.
```php
echo Formatter::toRoman(116); // Result: MMXXII
echo Formatter::toEmoji('thumbsup'); // Result: 👍
```
The examples are just random examples of what could be done (not saying we actually should).

**Reply by nibra** [reply]
> Let's define scenarios for the use of such objects.

```php
// Let $field be an array resulting from reading a forms.xml file
$showTitle = new Boolean($field['show_title']); // the string value is either 'show' or 'hide', mapping to true and false, rsp.
if ($showTitle.value === true) {
    echo 'Displaying the title is enabled';
}
$showTitle.value = false;
echo $showTitle; // Result: hide
$showTitle.value = 'on';
echo $showTitle; // Result: show
```

**Reply by korenevskiy** [reply]
Your example is impossible. Since it is unclear in which cases it is necessary to convert `$showTitle.value = true;` to the `show` or to `on` or `enabled`. 
It's more convenient this way.
```PHP
$showTitle = new Boolean($field['show_title']); // the string value is either 'show' or 'hide', mapping to true and false, rsp.
if ($showTitle == true) {
    echo 'Displaying the title is enabled';
}
$showTitle.value = false;
echo $showTitle.value; // Result: false
$showTitle.value = 'on';
echo $showTitle.value; // Result: on
```
`$showTitle = new Boolean($field['show_title'], [1=>'show',0=>'hide']);`
As you can see in the constructor, we specify the 2nd parameter that defines the conversion rule.
To be continued.
```PHP
$showTitle = new Stringean('This formatin text with any parameters `%s` and  `%s`', [1=>'show',0=>'hide']);
echo $showTitle; // Result: This formatin text with any parameters `show` and  `hide`
```

**Reply by korenevskiy** [reply]
But then we don't create anything new. Other than just copying the code by renaming the classes.
All these classes are the same. The **Boolean** class and the **Stringean** class are the same in every sense, they just have different names.
If we create such classes for each type of scalar quantity, then all these objects will be the same, with different names. In this case, we can just have a class with the general name Cast.

**Reply by korenevskiy** [reply]
I'm at a dead end. The ideas are good and at the same time there is no understanding of how to do it better. With the Bool type, there is no understanding of what is primary, type casting, or whether it is primary to be able to format. Let's leave the Bool type alone. This type has pitfalls. Let's discuss other scalar types. Let's understand how best to do for types with Int, String, Float. After that we will make a classes generalization . And after we define these types, we will go back to the Bool type.

**Reply by korenevskiy** [reply]
`Formatter` methods are very interesting. But these methods are imperative.
```PHP
JText::_('{ROMAN:116}');
JText::_('{EMOJI:thumbsup}');
```
We also know that emojis are localized images. How they will be displayed to people in a different latitude locality.
Here we can find out that the INI localization file may be the solution. But I have not studied this issue. How it works in the browser and how the browser adapts the emoji I do not know.

**Reply by korenevskiy** [reply]
Maybe it makes sense to modify the `JText` class? Maybe we could use it as a `Stringean` class?
```PHP
$showTitle = new JText('This formatin text with any parameters `%s` and  `%s`', [1=>'show',0=>'hide']);
echo $showTitle; // Result: This formatin text with any parameters `show` and  `hide`
// and
$showTitle = new JText([1=>'show',0=>'hide'], $field['show_title']);
$showTitle.value = true;
echo $showTitle; // Result: show
$showTitle.value = false;
echo $showTitle; // Result: hide
```

**Reply by korenevskiy** [reply]
continuation of the **`JText`** theme
```PHP
$showTitle = new JText([1=>'show',0=>'hide'], $field['show_title'], 'bool');
$showTitle.value = true; 
$showTitle.valueBool = "enabled"; 
```
the `valueBool` property will additionally convert to the Bool type

**Reply by korenevskiy** [reply]
This behavior will allow, at the same time, to convert characters to ROMAN. After that, substitute the result in the formatting string.
```PHP
//Typesetting Roman and formatting
$showTitle = new JText('FIELD_OPTION_TITLE', '116','roman');
//Typesetting Bool and formatting
$showTitle = new JText('FIELD_OPTION_TITLE', 'enable','bool');
//Typesetting Emoji and formatting
$showTitle = new JText('FIELD_OPTION_TITLE', 'thumbsup','emoji');
//Typesetting Emoji and Without formatting
$showTitle = new JText('', 'thumbsup','emoji');
$showTitle = new JText('', 'enable','bool');
//Typesetting Emoji and Without formatting
$showTitle = new JText('', 'thumbsup','roman');
```

**Reply by nibra** [reply]
> Your example is impossible. Since it is unclear in which cases it is necessary to convert `$showTitle.value = true;` to the `show` or to `on` or `enabled`.

It's not. The Boolean class knows the pairs of true/false values internally and memoizes which one to use at construction time. That's quite simple.

**Reply by nibra** [reply]
> If we create such classes for each type of scalar quantity, then all these objects will be the same, with different names. In this case, we can just have a class with the general name Cast.

A `String` class would contain (UTF-8 safe) methods to manipulate its value, f.x
```php
class String
{
    private string $value;

    public function contains(string $substring): bool
    {
        return mb_strpos($substring, $this->value) !== false;
    }

    public function length(): int
    {
        return mb_strlen($this->value);
    }

    // Other string functions
}
```

An `Integer` class would contain `increment()` and `decrement()` methods, and so on.

**Reply by nibra** [reply]
> How it works in the browser and how the browser adapts the emoji I do not know.

Those were just an exaple. However, these emojis are just UTF-8 characters.

**Reply by nibra** [reply]
> Maybe it makes sense to modify the `JText` class?

How would you do type hinting then?

**Reply by korenevskiy** [reply]
1. It is necessary to consider new types (Boolean, String, Float) from the point of view of typing for programming in the IDE. (Which IDE do you program in? (I use NetBeans)) 
2. In the examples above, we considered Boolean and String as classes that support string formatting. We want types with formatting. We dream of formatting-enabled types. Maybe we are mistaken? Maybe we need to do the analysis correctly, and just solve the problem with formatting of different types correctly. Metaphor: Similar to the fact that people often want to eat sweets, forgetting that they just need to eat regular food normally. We can also make a mistake about what we need to have.  In the C# programming language, we can create a new class by inheriting it from BOOL or from STRING. But there is no such thing in PHP, we do not know when such an opportunity will appear. We will not be able to compare the Boolean type simply with the BOOL type. The magic methods __conain(), __compare() and will not be for several years in PHP. If we create the types "Float", "Integer" they will not be used by anyone, since calling comparison functions from these types will strain and not give advantages in readability and understanding. We are already moving away from the JObject type. I propose to concentrate on the possibilities of formatting strings, perhaps by expanding the formatting of strings, the need for typing types will disappear. 
```PHP
/**
* @param string|array $format A string from the INI or an array to replace the value by the key.
* @param mixed $value  The source of the value
* @param string $type Type to cast the types of the $value attribute; The conversion method.
*  		$type possible values: emoji, roman, bool, int, string(default), float . . .
**/
class JText($format = '', $value = null, $type = ''){
}

// Initialization methods.
$text = new JText()
// or
$text = JText::getInstanse()

// Setting the value
$text->value  = 123;
// Setting the type
$text->type  = 'bool';
$text->type  = 'int';
$text->type  = 'float';
$text->type  = 'emoji';
$text->type  = 'roman';

// But if the property value was not set, then the property value will be used defined for this $type
$text->format= '';
$text->format= [true="show", false="hide"];

// The value is in its original state.
$text->source = $element['label'];

// The value will be converted when assigned to the desired type. By default to a string.
$text->value = $element['label'];
// But the source property will still have the value of the original state.
$text->source = $element['label'];

// The property is read-only.
echo $text->length;

// Supported methods: $haystack, $needle
$text->conaint($needle);
$text->start_with($needle);
$text->ends_with($needle);
$text->pad($length, $pad_string = " ");
$text->replace($search, $replace);
$text->split($string);
$text->ireplace($search, $replace);
$text->toupper();
$text->tolower();
$text->pos($needle);
$text->substring($start, $end);
// and
$text->_();
$text->sprintf($args);


// Print.
echo $text->value;
//Equivalent to
echo $text;
//Equivalent to
echo $text->_();
//Equivalent to
echo $text->sprintf($args);


// A short form of formatting.
echo JText::getInstanse('', 'thumbsup','emoji');	// Result: 👍
echo JText::getInstanse('', 'enable','bool');		// Result: 'true'
echo JText::getInstanse('', 116, 'roman');		// Result: MMXXII
echo JText::getInstanse('', 'smile', 'emoji');		// Result: 😀

// Enumeration of supported types.
JText::type_bool;
JText::type_roman;
JText::type_emoji;

//Support for objects and arrays
$user = ['name' => 'Agata', 'mail' => 'agata@mail.ru', 'phone' => '+79004881200', ];

// example 1
echo JText::getInstanse('My name %s, my mail %s, my phone %s', $user); 
// or
echo JText::getInstanse('FIELD_OPTIONS_TITLE', $user); 

// example 2
$text->format = 'My name %s, my mail %s, my phone %s';
$text->value = $user; 
echo $text;

// Result: 'My name Agata, my mail agata@mail.ru, my phone +79004881200'

// verify type
$text->type = 'int'; 
$text->value = '123'; 
echo gettype($text->value); 	// Retrun: INTEGER
echo gettype($text); 		// Retrun: STRING
```
I'm not good at the intricacies of formatting with parameters. But these are all the details of the work.
```PHP
$text->format = "My number: %s";
$text->type = 'integer';
$text->value = "123";
// But 
$text->value == 123; // Return: FALSE; // This is wrong! It should work differently.
// We will make sure that these values are equal!
$text->value == 123; // Return: TRUE;
// Therefore, the string format will be executed only when **`__toString()`** is called;
echo $text; // Return: "My number: 123"	// it is as STRING type
echo $text->value; // Return: 123 	// it is as INT type
```

**Reply by nibra** [reply]
How will that make dealing with the original problem easier?
(assume `$field['show_title']` contains 'show')
```php
$showTitle = new JText();
$showTitle->type  = 'bool';
$showTitle->format = [true="show", false="hide"];
$showTitle->value = $field['show_title'];
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```
vs.
```php
$showTitle = $field['show_title'] === 'show';
echo $showTitle ? 'show' : 'hide'; // Result: 'show'
$showTitle = !$showTitle;
echo $showTitle ? 'show' : 'hide'; // Result: 'hide
```
vs.
```php
$showTitle = new Boolean($field['show_title']);
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```

**Reply by korenevskiy** [reply]
> ```PHP
> $showTitle = new Boolean($field['show_title']);
> echo $showTitle; // Result: 'show'
> $showTitle->negate();
> echo $showTitle; // Result: 'hide'
> ```

You are right only when a boolean value can be represented as **`show`** & **`hide`**. 
But besides **`show`**&**`hide`**, there are other interpretations, It can be **`On`** & **`Off`** or **`Enable`** & **`Disable`** or **`Yes`** & **`No`** or **`Ok`** & **`None`**. How does this class when printing choose the correct display method?

If the attribute has the name $field['display'], then it is **`Show`**&**`Hide`**, but if the attribute has the name Enable, then it is **`On`**&**`Off`**. The Boolean type does not have a specific interpretation, the Boolean class cannot be applied without specifying the format for it, how it should be displayed when printing. In your Boolean class, you need to specify the parameter **`$showTitle->format = [true="show", false="hide"];`** just as it is specified in my example.

```PHP
$showTitle = new Boolean([true="show", false="hide"], $field['show_title']);
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```
or
```PHP
$showTitle = new Boolean( $field['show_title']);
$showtitle->format = [true="show", false="hide"];
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```
As a result, we will get the same thing, but just with a different name in the class name.

**Reply by korenevskiy** [reply]
```PHP
// this format can be defined as a constant in the JText class;
JText::boolYesNo 	// as [true="yes", false="no"]
JText::boolShowHide 	// as [true="show", false="hide"]
JText::boolOnOff 	// as [true="on", false="off"]

$showTitle = new JText(JText::boolShowHide, $field['show_title'],'bool');
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```
In this case, the type name is specified as a parameter in the constructor. But the rest of the functionality will be the same.

No matter how the class is called, we cannot avoid the format property. For any class name Boolean, Float, String, Integer, the format property will be a mandatory part. I want to explain to you that no matter how the class is called, the formatting of the object of this class will be the basis. I separated this formatting capability from the class object itself (Boolean,String,Float,Integer), and described it in the JText class; After that, it became clear to me that we needed not these Boolean, String, Float, Integer classes, but the ability to format them in a universal form with extensive capabilities and support for a short form of writing.

**Reply by nibra** [reply]
Your example
```php
$showTitle = new JText(JText::boolShowHide, $field['show_title'],'bool');
echo $showTitle; // Result: 'show'
$showTitle->negate();
echo $showTitle; // Result: 'hide'
```
will not work, as `JText` does not have a `negate()` method. A method with that name only makes sense for booleans, not for text.

**Comment by nibra** [comment]
> How does this class when printing choose the correct display method?

It internally knows the valid true/false pairs and decides which one to choose depending on the value provided to the constructor.
```php
$var = new Boolean('show');
echo $var; // Result: 'show'
$var->negate();
echo $var; // Result: 'hide'
```
```php
$var = new Boolean('on');
echo $var; // Result: 'on'
$var->negate();
echo $var; // Result: 'off'
```
```php
$var = new Boolean('yes');
echo $var; // Result: 'yes'
$var->negate();
echo $var; // Result: 'no'
```
and so on.

**Reply by nibra** [reply]
> After that, it became clear to me that we needed not these Boolean, String, Float, Integer classes, but the ability to format them in a universal form with extensive capabilities and support for a short form of writing.

In that case, I see no benefit over the native PHP way to solve that:
```php
$showTitle = $field['show_title'] === 'show';
echo $showTitle ? 'show' : 'hide'; // Result: 'show'
$showTitle = !$showTitle;
echo $showTitle ? 'show' : 'hide'; // Result: 'hide
```

Although the Boolean class does not require significantly less code, the benefit is that you don't need to know that `$field['show_title']` encodes `true` and `false` as 'show' and 'hide'. Anything being more complex than the native PHP way does not make sense, because nobody will ever use it.

**Comment by korenevskiy** [comment]
```PHP
$var = $field['show_title'];
echo $var; // Result: 'yes'
boolean_negate($var);
echo $var; // Result: 'no'
```
or 
```PHP
$var = 'show';
echo $var; // Result: 'yes'
boolean_negate($var);
echo $var; // Result: 'no'
```
```PHP
/ * *
* @param mixed  $var 
* @return string
*/
function boolean_negate($var) : string{
}
```
an easy-to-understand script, easy to read.
or 
```PHP
// Static methods
echo boolean::negate($var);
```

**Comment by korenevskiy** [comment]
@nibra 
Do you think this equality should work like this?
```PHP
echo ((new Boolean('yes')) != (new Boolean('on')));
```

**Reply by nibra** [reply]
That's not possible, because PHP does not support operator override (yet). It would be
```php
(new Boolean('yes')).value !== (new Boolean('no')).value
```

**Reply by korenevskiy** [reply]
I want to say that `__toString()` should not return "yes" or "on". since they will not be equal when compared. It is for this reason that there is no point in the `Boolean` object. Therefore, only static methods can be used to implement your idea.

**Reply by nibra** [reply]
In that case, there is no point in this idea at all. As shown above, all other approaches produce more overhead and cognitive load.

**Reply by korenevskiy** [reply]
Let's clarify. I offered several ideas in this discussion. each of my ideas are designed for completely different applications. The fact is that at first I did not understand how your idea should work.
It is precisely because the Boolean object creates a cognitive load that I suggested making the methods static. So that the Boolean class is not an object, but is just a helper for static methods.
The need you described to invert the value of "YES" to the opposite, I consider a very useful function.
For the Boolean class, we will also create a static method that will allow you to compare 2 string values in different formats.
```PHP
echo Boolean::compare('show','on') ? 'True' :  'False';
```
I suggest making an class for static methods
```PHP
class Boolean{
	public static function compare($first = "" , $second = ""): bool{
	}
	public static function negate($value = "") : string{
	}
	public static function is($value = "") : bool{
	}
}
```
What folder do you think this class can be placed in?

**Reply by korenevskiy** [reply]
What do you think are the disadvantages in this example?

**Reply by nibra** [reply]
> What do you think are the disadvantages in this example?

I don't see any, but I don't see any significant advantage either. You still have to deal with strings in the application that originally just belong in the transportation (storage) layer. Since casting from and to types needed by that layer is the duty of the model, it coud as well make real booleans from the provided values and vice versa. That would be the most straight forward solution, since the application does not need to know how the boolean values where stored.

**Reply by korenevskiy** [reply]
I love typization, I use typization in the IDE for error checking. I have a lot of experience with C#. But the bottom line is that if we cannot compare 2 objects with the same type, then there is no point in this typization. Such typization will be misleading. 
Exactly 1 year later, I will start using an object of the Boolean class, I will forget how it was created, of course, when using this object, I will compare it with each other. I will get a lot of errors, because in this case typization will not simplify the code, but complicate it, since I will not be able to compare objects with each other.

**Reply by korenevskiy** [reply]
> That would be the most straight forward solution, since the application does not need to know how the boolean values where stored.

In this case, the application can use a variable in `Bool` format or in `String` format as the original value. In the case when it is necessary to perform checks, we can use the static assistant function.

I believe that an object of the Boolean class should be created only if objects of this type can be compared with each other.  If they cannot be compared, then we need to make do with static helpers, it will be typization with native types, at least it will not create problems.

**Reply by korenevskiy** [reply]
We can do this: 
```PHP
$var = new Boolean('yes');
echo $var; 		// Return: "true"
echo $var->value; 	// Return: "yes"
echo $var->negate();	// Return: "no"
echo $var == (new Boolean("on")) ? "True" : "False";	// Return: "True"
```

**Reply by nibra** [reply]
That would require to implement all boolean operators in the Boolean class.

**Reply by korenevskiy** [reply]
Of course we need to do this. We can use an array. Find the logical value by using the string key.

**Reply by korenevskiy** [reply]
@nibra 
It is also necessary to achieve so that the object can be compared with the usual Bool value.
```PHP
$var = new Boolean('yes'); 
echo $var == true ? "True" : "False";	// Return: "True"
$var = new Boolean('no'); 
echo $var == false ? "False" : "True";	// Return: "False"
```
This is possible?

**Reply by nibra** [reply]
Not with PHP, until it will provide [operator overloads](https://wiki.php.net/rfc/userspace_operator_overloading), which not seems to happen.

**Reply by korenevskiy** [reply]
> Not with PHP, until it will provide [operator overloads](https://wiki.php.net/rfc/userspace_operator_overloading), which not seems to happen.

I know that rebooting in PHP does not exist. I wanted to offer an idea.
Can the `__toString()` method return the desired value so that it can meet the condition above?

**Reply by korenevskiy** [reply]
@nibra 
I checked, such a comparison is not possible in PHP .
Let's ask other participants how to recommend creating comparisons of stringed instin values and how to implement this.
